### DISCUSSION TITLE
best way to use access token to call the external API

### DISCUSSION THREAD

**Discussion by sursaone** [gladly-team/next-firebase-auth#180] [Q&A]
I have APIs exported as functions from common lib.
e.g. lib/services/teams.ts
```ts
export const getMyTeam = ()=> {
  const token =`??`
  fetch(`${BASE_URL}/api/teams`, {
     method: 'GET',
     headers: {
        authorization: `Bearer ${token}`,
        'Content-Type': 'application/json'
     }
 }
}
```
How can I get the token in above case so that same function will be working with both SSR and SSG?
Note: I am using the react-query so that I can't pass the token id or current user to the function method.

Also, how can I use this with another provider? e.g. I want to use logrocket with the custom provider and to set current user.

**Comment by kmjennison** [comment]
To get the user's ID token, use [`useAuthUser`](https://github.com/gladly-team/next-firebase-auth#useauthuser). There's an [example app here](https://github.com/gladly-team/next-firebase-auth/tree/main/example) that has a [live demo](https://github.com/gladly-team/next-firebase-auth#demo).

The `useAuthUser` is a hook—so for the above example, you'll probably want your module to take the token as an argument, or refactor the fetch/authorization into its own module that manages  the token value.

**Reply by sursaone** [reply]
@kmjennison 
> refactor the fetch/authorization into its own module that manages the token value.

Any sample for this would be helpful, as useAuthUser is the hook and can't be used outside and without withAuthUser HOC. How can I achieve this?

**Reply by shreyas-jadhav** [reply]
Have you tried using the Firebase SDK directly?

```js
firebase.auth().currentUser.getIdToken(/* forceRefresh */ true).then(function(idToken) {
  // Send token to your backend via HTTPS
  // ...
}).catch(function(error) {
  // Handle error
});
```

**Reply by sursaone** [reply]
@shreyas-jadhav But it won't work on SSR, right?

**Reply by shreyas-jadhav** [reply]
Yup, not for SSR obviously. My bad.

Let me know if this helps.

```js

export​ ​const​ ​getServerSideProps​ ​=​ ​withAuthUserTokenSSR​(​{​
  ​whenUnauthed​: ​AuthAction​.​REDIRECT_TO_LOGIN​,​
​}​)​(​async​ ​(​{​ AuthUser ​}​)​ ​=>​ ​{​
  ​// Optionally, get other props.​
  ​const​ ​token​ ​=​ ​await​ ​AuthUser​.​getIdToken​(​)​
  ​const​ ​response​ ​=​ ​await​ ​fetch​(​'/api/my-endpoint'​,​ ​{​
    ​method​: ​'GET'​,​
    ​headers​: ​{​
      ​Authorization​: ​token​,​
    ​}​,​
  ​}​)​
  ​const​ ​data​ ​=​ ​await​ ​response​.​json​(​)​
  ​return​ ​{​
    ​props​: ​{​
      ​thing​: ​data​.​thing​
    ​}​
  ​}​
​}​)​

```

**Reply by sursaone** [reply]
@shreyas-jadhav This I saw but I am using reusable APIs that are on separate .ts files. And I want to access this token from a single point where I am configuring the fetch headers, like axios request interceptors.

**Reply by kmjennison** [reply]
@sursaone [Here's](https://github.com/gladly-team/tab-web/blob/master/src/utils/hooks/useData.js) how we do this in one of our apps. That code is made more complicated because we use Relay/GraphQL, but the basic idea is:
1. [Create/recreate a fetcher function](https://github.com/gladly-team/tab-web/blob/adfd8a6a4198f5ee89eee38b7def069a83411596/src/utils/relayEnvironment.js#L29) when the AuthUser changes, setting the Authorization header.
2. Use that fetcher to call APIs, both during SSR and [from hooks](https://github.com/gladly-team/tab-web/blob/master/src/utils/hooks/useData.js#L11).

You might want to check out [SWR](https://github.com/vercel/swr), which I find helpful to think about how to handle data fetching with hooks.

I agree, it would be good to add a simple example of this to the documentation.

**Reply by yeuem1vannam** [reply]
@kmjennison 
I'm a bit worried because the `idToken` is [not recommended to use in API call](https://auth0.com/blog/id-token-access-token-what-is-the-difference/#:~:text=One%20of%20the%20most%20common%20mistakes%20developers%20make%20with%20an%20ID%20token%20is%20using%20it%20to%20call%20an%20API.)

If there is any way to exchange the `idToken` (of Firebase) to API's `accessToken` during the login procedure, it should be a great.

**Reply by kmjennison** [reply]
@yeuem1vannam Can you link to Firebase docs that discuss using an access token?
